clang/test/CodeGenSYCL/sycl-intelfpga-annotated-field.cpp

```diff
@@ -0,0 +1,78 @@
+// RUN: %clang_cc1 -no-opaque-pointers -fsycl-is-device -fintelfpga -triple spir64_fpga -aux-triple x86_64-unknown-linux-gnu -IInputs -emit-llvm  -o - %s | FileCheck %s
+
+// This test checks that clang emits @llvm.ptr.annotation intrinsic correctly
+// when attribute annotate is applied to a kernel functor field.
+
+// CHECK: @[[STR1:.*]] = private unnamed_addr addrspace(1) constant [9 x i8] c"my_ann_1\00", section "llvm.metadata"
+// CHECK: @[[STR2:.*]] = private unnamed_addr addrspace(1) constant [9 x i8] c"my_ann_2\00", section "llvm.metadata"
+// CHECK: @[[STR3:.*]] = private unnamed_addr addrspace(1) constant [9 x i8] c"my_ann_3\00", section "llvm.metadata"
+
+#include "Inputs/sycl.hpp"
+
+struct bar {
+  _BitInt(5) a;
+};
+
+struct fooA {
+    [[clang::annotate("my_ann_1")]]
+    int *p;
+
+    fooA(int *_p) : p(_p) {}
+    void operator()() const {}
+};
+
+struct fooB {
+    [[clang::annotate("my_ann_2")]]
+    float f;
+
+    fooB(float _f) : f(_f) {}
+    void operator()() const {}
+};
+
+struct fooC {
+    [[clang::annotate("my_ann_3")]]
+    bar b;
+
+    fooC(bar _b) : b(_b) {}
+    void operator()() const {}
+};
+
+int main() {
+  sycl::handler h;
+  h.single_task(fooA{nullptr});
+  h.single_task(fooB{2.0});
+  h.single_task(fooC{{3}});
+  return 0;
+}
+
+// check IR pattern for int* kernel field
+// CHECK: define dso_local spir_kernel void @_ZTS4fooA(i32 addrspace(1)* noundef align 4 %_arg_p) #0
+// CHECK: %_arg_p.addr = alloca i32 addrspace(1)*, align 8
+// CHECK: %_arg_p.addr.ascast = addrspacecast i32 addrspace(1)** %_arg_p.addr to i32 addrspace(1)* addrspace(4)*
+// CHECK: store i32 addrspace(1)* %_arg_p, i32 addrspace(1)* addrspace(4)* %_arg_p.addr.ascast, align 8
+// CHECK: %1 = call i8 addrspace(4)* @llvm.ptr.annotation.p4i8.p1i8({{.*}}@[[STR1:.*]],
+// CHECK: %2 = bitcast i8 addrspace(4)* %1 to i32 addrspace(4)* addrspace(4)*
+// CHECK: %3 = load i32 addrspace(1)*, i32 addrspace(1)* addrspace(4)* %_arg_p.addr.ascast, align 8
+// CHECK: %4 = addrspacecast i32 addrspace(1)* %3 to i32 addrspace(4)*
+// CHECK: store i32 addrspace(4)* %4, i32 addrspace(4)* addrspace(4)* %2, align 8
```

Comment:
What exactly does the IR pattern matcher attempt to look at?   This seems a little fragile to me.
Also, please add a opaque-ptr-version of the same test.

Reply:
More details about the targeted pattern is added in the description.

opaque-ptr version of the new lit test is added.